File: src/LinkedList/LC206ReverseLinkedList.java
package LinkedList;

public class LC206ReverseLinkedList {

    public ListNode reverseList(ListNode head) {
        ListNode current = head;
        ListNode nextNode = head == null ? null : head.next;
        ListNode previousNode = null;

        while (current != null) {
            // reverse the link
            current.next = previousNode;

            previousNode = current;
            current = nextNode;
            nextNode = nextNode != null ? nextNode.next : null;
        }

        return previousNode;
    }

    public static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}

File: src/LinkedList/LC206ReverseLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  if (!head) return null;

  let prev = null;
  let curr = head;
  let currNext = curr.next ? curr.next : null;

  while (curr) {
    // reverse link
    curr.next = prev;

    // move one node forward
    prev = curr;
    curr = currNext;
    currNext = currNext ? currNext.next : null;
  }

  return prev;
};

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}
